### ISSUE TITLE
Missing "Dhivehi" Language. 

### ISSUE THREAD

**Issue by Xayaan** [tesseract-ocr/tessdata#43]
Tessdata had Dhivehi language but its missing now.

**Comment by amitdo** [comment]
https://github.com/tesseract-ocr/tesseract/blob/9c7e99b04197fb9900c29be8bb9ac79a7a8b4672/training/language-specific.sh#L32

**Comment by amitdo** [comment]
div.traineddata was added to the repo
https://github.com/tesseract-ocr/tessdata/blob/master/best/div.traineddata

**Comment by Shreeshrii** [comment]
@Xayaan Have you tried the 'best' version? Any feedback?

**Comment by Xayaan** [comment]
@Shreeshrii yes i have tried and the results are pretty bad. It needs to be trained intensively.

**Comment by Shreeshrii** [comment]
Do not close the issue then, you can change title to say feedback regarding Dhivehi and then add some notes regarding what is wrong so that it can be improved.

**Comment by Xayaan** [comment]
Done, thank you! 👍

**Comment by Shreeshrii** [comment]
Are these fonts suitable for Dhivehi ?

http://www.hassanhameed.com/?page_id=152

**Comment by Xayaan** [comment]
Yes, but I'd trust these : https://dhivehi.mv/fonts/

**Comment by Shreeshrii** [comment]
https://dv.wikipedia.org/wiki/%DE%89%DE%A6%DE%87%DE%A8_%DE%9E%DE%A6%DE%8A%DE%B0%DE%99%DE%A7

@theraysmith 
This script looks similar to Arabic with accents. Have you had success in adding the accented version for next training?

**Comment by amitdo** [comment]
There is also Thaana traineddata

**Comment by Shreeshrii** [comment]
@Xayaan please check with Thaana traineddata also.

If possible, provide an image and it's corresponding ground truth file for testing.

**Comment by Xayaan** [comment]
Yes, it is. It is similiar to sanskrit and arabic. A RTL language. 

I checked with the thaana traineddata, its not very accurate and has low accuracy now. 

@Shreeshrii Would you like to work with me and make the OCR accurate to dhivehi? There is business opportunity here.

**Comment by Sofwath** [comment]
any pointers to the training data used for Thaana?

**Comment by Shreeshrii** [comment]
The langdata repo has not been updated for 4.0x

https://github.com/tesseract-ocr/langdata/tree/master/tha
has the old training files

ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by Sofwath** [comment]
it looks like that is for Thai. How about for Thaana (div)

**Comment by Shreeshrii** [comment]
Sorry about that.

Looks like https://github.com/tesseract-ocr/langdata/tree/master/div
does not have all the required files.

If it is similar to Arabic, you can copy langdata files from there and
modify for Thaana.

http://crubadan.org/languages/dv
could be a source for wordlists, training text.

ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by Sofwath** [comment]
Link to Thaana (div) monogram and bigram file 

https://github.com/Sofwath/thaanaOCR/tree/master/data

This is a Thaana text corpus 

https://www.dropbox.com/s/04ox44rfuqm5xhw/dv_MV_1.txt?dl=0

Anything else that we need to have for a basic training ?

**Comment by Shreeshrii** [comment]
You can download

https://github.com/tesseract-ocr/tessdata_best/blob/master/div.traineddata
and
https://github.com/tesseract-ocr/tessdata_best/blob/master/Thaana.traineddata

Then unpack the traineddata to get the files.

root@All-in-1-Touch:/mnt/c/Users/User/shree/tessdata_best# combine_tessdata
-u div.traineddata div.
Extracting tessdata components from div.traineddata
Wrote div.lstm
Wrote div.lstm-punc-dawg
Wrote div.lstm-word-dawg
Wrote div.lstm-number-dawg
Wrote div.lstm-unicharset
Wrote div.lstm-recoder
Wrote div.version
Version string:4.00.00alpha:div:synth20170629
17:lstm:size=3218139, offset=192
18:lstm-punc-dawg:size=4506, offset=3218331
19:lstm-word-dawg:size=1342450, offset=3222837
20:lstm-number-dawg:size=426, offset=4565287
21:lstm-unicharset:size=7276, offset=4565713
22:lstm-recoder:size=1093, offset=4572989
23:version:size=30, offset=4574082

root@All-in-1-Touch:/mnt/c/Users/User/shree/tessdata_best# combine_tessdata
-u Thaana.traineddata Thaana.
Extracting tessdata components from Thaana.traineddata
Wrote Thaana.lstm
Wrote Thaana.lstm-punc-dawg
Wrote Thaana.lstm-word-dawg
Wrote Thaana.lstm-number-dawg
Wrote Thaana.lstm-unicharset
Wrote Thaana.lstm-recoder
Wrote Thaana.version
Version string:4.00.00alpha:Thaana:synth20170629
17:lstm:size=7723707, offset=192
18:lstm-punc-dawg:size=5674, offset=7723899
19:lstm-word-dawg:size=5036906, offset=7729573
20:lstm-number-dawg:size=4762, offset=12766479
21:lstm-unicharset:size=10741, offset=12771241
22:lstm-recoder:size=1633, offset=12781982
23:version:size=33, offset=12783615
root@All-in-1-Touch:/mnt/c/Users/User/shree/tessdata_best#


You can further get the original wordlists by using
dawg2wordlist

But the actual training_text will not be there.

ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by Sofwath** [comment]
Great. Thanks. Will work on that.

**Comment by Shreeshrii** [comment]
dawg2wordlist syntax will be as follows

$ dawg2wordlist Thaana.lstm-unicharset Thaana.lstm-word-dawg Thaana.wordlist
Loading word list from Thaana.lstm-word-dawg
Reading squished dawg
Word list loaded.

similarly for punc and numbers.

You can review these files for accuracy.

I don't think tesseract uses unigrams and bigrams for training, though they
maybe used internally at Google to generate a representative training text.


ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by Shreeshrii** [comment]
FYI Thaana files will have both English and Divehi. div files will have
only Divehi.



ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by Sofwath** [comment]
question: do i still need to create the box files even if we are using the lstm method?

**Comment by Shreeshrii** [comment]
You have to use tesstrain.sh script file, also see tesstrain_utils.sh and
language_specific.sh in training directory.

These create the box/tiff files from the training text and specified fonts.
They are used for creating the lstmf files and are kept only in the tmp
directory.

Try the training tutorial for english and look at the log file and tmp
directory.

**Comment by Sofwath** [comment]
Any help? I am getting this error

sudo /Users/sofwath/tesseract/training/tesstrain.sh --fonts_dir /Users/sofwath/dev/MLAI/tesseract/font/ --lang div --linedata_only --noextract_font_properties --langdata_dir langdata --tessdata_dir tessdata/ --output_dir divtrain/

=== Starting training for language 'div'
mktemp: illegal option -- -
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix 
[Wed Oct 4 10:49:19 +05 2017] /usr/local/bin/text2image --fonts_dir=/Users/sofwath/dev/MLAI/tesseract/font/ --font=MV Typewriter --outputbase=/sample_text.txt --text=/sample_text.txt --fontconfig_tmpdir=

=== Phase I: Generating training images ===
Rendering using MV Typewriter
[Wed Oct 4 10:49:20 +05 2017] /usr/local/bin/text2image --fontconfig_tmpdir= --fonts_dir=/Users/sofwath/dev/MLAI/tesseract/font/ --strip_unrenderable_words --leading=32 --char_spacing=0.0 --exposure=0 --outputbase=/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.GrmD2E1v/div/div.MV_Typewriter.exp0 --max_pages=3 --font=MV Typewriter --text=langdata/div/div.training_text
ERROR: /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.GrmD2E1v/div/div.MV_Typewriter.exp0.box does not exist or is not readable
ERROR: /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.GrmD2E1v/div/div.MV_Typewriter.exp0.box does not exist or is not readable

**Comment by Shreeshrii** [comment]
You are getting error

mktemp: illegal option -- -
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
mktemp [-d] [-q] [-u] -t prefix

------------

see tesstrain_utils.sh
lines 29 and 172

training uses the /tmp directory for creating files



ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by Sofwath** [comment]
changed 
    export FONT_CONFIG_CACHE=$(mktemp -d --tmpdir font_tmp.XXXXXXXXXX)
to
     export FONT_CONFIG_CACHE=$(mktemp -d -tmpdir font_tmp.XXXXXXXXXX)
in tesstrain_utils.sh and the first error was fixed but still get 

=== Starting training for language 'div'
/Users/sofwath/tesseract/training/tesstrain_utils.sh: line 197: ${sample_path}: ambiguous redirect
[Wed Oct 4 14:05:34 +05 2017] /usr/local/bin/text2image --fonts_dir=/Users/sofwath/dev/MLAI/tesseract/font/ --font=MV Typewriter --outputbase=/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/mpdir.8n1I8qtq font_tmp.f7jJ5RoYKx/sample_text.txt --text=/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/mpdir.8n1I8qtq font_tmp.f7jJ5RoYKx/sample_text.txt --fontconfig_tmpdir=/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/mpdir.8n1I8qtq font_tmp.f7jJ5RoYKx
'--text' option is missing!

=== Phase I: Generating training images ===
Rendering using MV Typewriter
[Wed Oct 4 14:05:34 +05 2017] /usr/local/bin/text2image --fontconfig_tmpdir=/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/mpdir.8n1I8qtq font_tmp.f7jJ5RoYKx --fonts_dir=/Users/sofwath/dev/MLAI/tesseract/font/ --strip_unrenderable_words --leading=32 --char_spacing=0.0 --exposure=0 --outputbase=/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.OaGHUsk9/div/div.MV_Typewriter.exp0 --max_pages=3 --font=MV Typewriter --text=langdata/div/div.training_text
'--text' option is missing!
ERROR: /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.OaGHUsk9/div/div.MV_Typewriter.exp0.box does not exist or is not readable
ERROR: /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.OaGHUsk9/div/div.MV_Typewriter.exp0.box does not exist or is not readable

**Comment by Shreeshrii** [comment]
You have to look at your paths

'--text' option is missing!

ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by Sofwath** [comment]
I followed this example

training/tesstrain.sh --fonts_dir /usr/share/fonts --lang eng --linedata_only \
  --noextract_font_properties --langdata_dir ../langdata \
  --tessdata_dir ./tessdata --output_dir ~/tesstutorial/engtrain

**Comment by Sofwath** [comment]
for tesstrain.sh there is no --text command line option

**Comment by Shreeshrii** [comment]
/Users/sofwath/tesseract/training/tesstrain_utils.sh: line 197:
${sample_path}: ambiguous redirect

If you are changing the bash script, you have to make sure it is done
correctly. Please look at the error messages you get.

ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by Sofwath** [comment]
I've been trying on Mac OS . giving too many errors on the bash scripts. I'll try to run the process on Linux

**Comment by nashrafeeg** [comment]
@Sofwath is the dhivehi training data usable  now or is it abandoned? i have not seen any updates regarding this since 2017
